README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# starsTileServer

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

The goal of starsTiles is to provide a simple way to add large data sets
in R to a leaflet application. Examples are available on the
[website](https://bartk.gitlab.io/starsTileServer/). Because it is
possible to provide functions pretty much any data that can be projected
can be visualized, for example composites or non square grids. This is
en example of using data from meteorological radar:

<img src="man/figures/composite.png" width="48%" /><img src="man/figures/vrad.png" width="48%" />

## Installation

You can install the development version as follows:

``` r
remotes::install_gitlab('bartk/starsTileServer')
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(starsTileServer)
```

Some example data can be downloaded as follows

``` r
require(ecmwfr)
request <-
  list(
    "dataset_short_name"  = "reanalysis-era5-pressure-levels",
    "product_type"  = "reanalysis",
    "variable"       = c("temperature", "geopotential", "u_component_of_wind", "v_component_of_wind"),
    "pressure_level" = c("875", "900", "925"),
    "year"           = "2000",
    "month"          = "04",
    "day"            = as.character(27:29),
    "time"           = sprintf("%02i:00", 0:23),
    "area"           = "64/-130/-64/144",
    "format"         = "netcdf",
    "target"         = "test.nc"
  )
# make sure you use your own uid and key ( https://cds.climate.copernicus.eu/#!/home )
wf_set_key("uid", "key", service = "cds")
ncfile <- wf_request(
  user = "uid",
  request = request,
  transfer = TRUE,
  path = "~",
  verbose = FALSE
)
```

### Setting up the server

The following commands would start a tile server

``` r
# note the process is ran in the background, do not forget to close it as it might use quite a bit of memory.
require(callr)
#> Loading required package: callr
rp <- r_bg(function() {
  # read a stars grid
  weatherData <- stars::read_stars("~/weatherData.nc", proxy = FALSE, sub = "t")
  names(weatherData) <- "t"
  sf::st_crs(weatherData) <- "+proj=longlat"
  colorFunction <- leaflet::colorNumeric("viridis", c(250, 310))
  colorFunctionWithAlpa <- function(x, alpha = 1) {
    paste0(colorFunction(x), as.character(as.raw(
      as.numeric(alpha) * 255
    )))
  }
  starsTileServer::starsTileServer$new(weatherData, colorFun = colorFunctionWithAlpa)$run(port = 3746)
})
```

### Visualizing the result

In a separate R session the data can be added to a leaflet map (in this
case temperature from the 900 mb level, at 2000-04-27 01:00:00)

``` r
require(leaflet)
#> Loading required package: leaflet
require(leaflet.extras)
#> Loading required package: leaflet.extras
map <- leaflet() %>%
  addTiles() %>%
  enableTileCaching() %>%
  addTiles(
    "http://127.0.0.1:3746/map/t/{z}/{x}/{y}?level=900&time=2000-04-27 01:00:00&alpha=0.5",
    options = tileOptions(useCache = TRUE, crossOrigin = TRUE)
  ) %>%
  setView(zoom = 3, lat = 30, lng = 30)
```

This map looks as follows:

``` r
map
```

<img src="man/figures/README-plot_map_image-1.png" width="100%" />

The documentation of the api is available by appending url with
`__docs__` in this case: `http://127.0.0.1:3746/__docs__/`.

We have to make sure we close the server that is running as a separate
process:

``` r
message(rp$read_output())
#> t,
message(rp$read_error())
#> Running plumber API at http://127.0.0.1:3746
#> Running swagger Docs at http://127.0.0.1:3746/__docs__/
rp$finalize()
```

# Alternative mapping solutions

Several other solutions do exist for visualizing gridded data on leaflet
maps, for example `leafem::addStarsImage`, these do however differ in
two distinct ways. First they render static grids, therefore visualizing
times series of data or dynamically switching to a different
layer/dataset might be quite challenging. Second the `starsTileServer`
makes zooming in the map possible. For each zoom level the map is
dynamically rendered meaning it is possible to zoom in and out without
loosing any information.

Another package that can be used for visualizing gridded data in
`leaflet` is `tiler`. This package renders a static set of tiles. The
advantage is that no server needs to be running to use these tiles.
However the disadvantage is that all rendering needs to be done
beforehand. For large datasets this can require considerable resources
both for storage and calculation.

Dynamic rendering opens also possibilities to visualize other data
dynamically. The package allows for custom functions that render tiles.
These can be used to visualize derived attributes (e.g. wind speed or
uplift), data from other data sources (like databases) or non-regular
data (e.g. radar images).

Another big difference is that the data storage and the computer where
it is visualized do not have to be the same for a tile server. One tile
server can be used by many users. It is also possible to scale the
rendering resources by making the rendering process parallel.

# FAQ

-   What are `tiles`?

`Tiles` are the small images used by online maps to visualize the maps.
These are generally `256*256` pixels and generated on various zoom
level. More info can be found on
[wikipedia](https://en.wikipedia.org/wiki/Tiled_web_map).


NEWS.md
# starsTileServer 0.1

* Release package
* Fix typos reported by Anne Scharf
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(starsTileServer)

test_check("starsTileServer")


tests/testthat/test-supportFun.R
test_that("coordinate identification", {
  expect_equal(lonfun(xfun(4.34, 5), 5), 4.34)
  expect_equal(xfun(0, 0), .5)
  expect_equal(xfun(-90, 1), .5)
  expect_equal(xfun(90, 1), 1.5)
  expect_equal(yfun(0, 0), .5)
  expect_equal(latfun(0, 0), latfun(0, 1))
  expect_equal(latfun(0, 0), latfun(0, 7))
  expect_equal(latfun(1, 0), latfun(2, 1))
  expect_equal(latfun(yfun(4.34, 5), 5), 4.34)
})

test_that("target matrix", {
  expect_equal((dim(targetGrid(4, 4, 3))), c(x = 256L, y = 256L))
  expect_false(any(sf::st_overlaps(sf::st_as_sf(starsTileServer:::targetGrid(3, 3, 5, 10)), sf::st_as_sf(starsTileServer:::targetGrid(3, 4, 5, 10)), F)))
  expect_equal(sum(sf::st_touches(sf::st_as_sf(starsTileServer:::targetGrid(3, 3, 5, 4)), sf::st_as_sf(starsTileServer:::targetGrid(3, 4, 5, 4)), F)), 10)
})

test_that("target matrix errors", {
  expect_error(targetGrid(-1, 0, 0))
  expect_error(targetGrid(0, -1, 0))
  expect_equal(class(targetGrid(0, 0, 0)), "stars")
  expect_error(targetGrid(0, 0, -1))
})


tests/testthat/helper-mock-request.R

# from : https://github.com/rstudio/plumber/blob/main/tests/testthat/helper-mock-request.R
make_req <- function(verb = "GET", path = "/", qs = "", body = "", args = c(), pr = NULL, ...) {
  req <- as.environment(list(...))
  req$REQUEST_METHOD <- toupper(verb)
  req$PATH_INFO <- path
  req$QUERY_STRING <- qs

  if (is.character(body)) {
    body <- charToRaw(body)
  }
  stopifnot(is.raw(body))
  req$rook.input <- list(
    read_lines = function() {
      rawToChar(body)
    },
    read = function() {
      body
    },
    rewind = function() {
      length(body)
    }
  )
  req$bodyRaw <- body
  req$pr <- pr
  req
}


R/supportFun.R
#' @importFrom grDevices col2rgb
#' @importFrom png writePNG
#' @importFrom units drop_units
#' @importFrom leaflet colorNumeric
#' @importFrom R6 R6Class
#' @importFrom assertthat assert_that is.count
#' @importFrom sf st_bbox st_transform st_as_sfc st_crs st_join st_drop_geometry st_as_sf
#' @import stars
#' @import plumber
#' @import rlang
NULL

xfun <- function(lon, z) {
  1 / (2 * pi) * 2^z * ((lon * pi / 180) + pi)
}
lonfun <- function(x, z) {
  ((x / (1 / (2 * pi) * 2^z)) - pi) / pi * 180
}
yfun <-
  function(lat, z) {
    1 / (2 * pi) * 2^z * (pi - log(tan(pi / 4 + (lat * pi / 180) / 2)))
  }
latfun <-
  function(y, z) {
    (atan(exp(pi - (y / (
      1 / (2 * pi) * 2^z
    )))) - pi / 4) * 2 * 180 / pi
  }

#' Function to calculate a stars grid based on the x,y,z attributes that represents the coordinates of the map tile
#'
#' @param x The location of the tile in the x dimension
#' @param y The location of the tile in the y dimension
#' @param z The zoom level
#' @param tileSize The size of the tile generally 256 pixels
#'
#' @return A stars grid with all values being zero with the dimensions matching those required for the tile specified
#'
#' @details
#'
#' This function is mostly useful for testing purpose.
#'
#' @export
#'
#' @examples
#' targetGrid(4, 2, 4)
#' targetGrid(4, 2, 4, 128)
targetGrid <- function(x, y, z, tileSize = 256L) {
  assert_that(is.count(x + 1L))
  assert_that(is.count(y + 1L))
  assert_that(is.count(z + 1L))
  assert_that(is.count(tileSize))

  targetExtent <-
    st_bbox(c(
      xmin = lonfun(x, z),
      xmax = lonfun(x + 1, z),
      ymin = latfun(y + 1, z),
      ymax = latfun(y, z)
    ),
    crs = 4326
    )
  st_as_stars(
    nx = tileSize,
    ny = tileSize,
    st_bbox(st_transform(
      st_as_sfc(targetExtent),
      crs = 3857 # web mercator
    ))
  )
}
# Converts a matrix of values to to a png image that it the tile to be returned
matrixToPng <- function(targetMatrix, colFun, ...) {
  t <- ncol(targetMatrix) # assums square matrix
  if (inherits(targetMatrix, "units")) {
    targetMatrix <- drop_units(targetMatrix)
  }
  colorArray <-
    aperm(array(col2rgb(
      ifelse(is.na(targetMatrix), "#00000000", colFun(targetMatrix, ...)),
      alpha = T
    ) / 255,
    dim = c(4, t, t)
    ), c(3, 2, 1))[, , ]
  writePNG(colorArray)
}

# Prevent warning from checks for arguments that are not in the function definition in the following two functions
utils::globalVariables(c("x", "y", "z", "colFun", "attrib", "private"))

handlerFunctionFun <- function() {
  args <- as.list(environment())[-(1:3)]
  # Find the grid for which the values need to be calculated
  targetGrid <- targetGrid(x, y, z, private$tileSize)
  nms <- names(args)[names(args) %in% names(formals(private$grid[["idName"]]))]
  targetGridWithValues <-
    do.call(private$grid[["idName"]], c(list(targetGrid), args[nms]))
  targetMatrix <- (drop(targetGridWithValues[[1]]))
  nms <- names(args)[names(args) %in% names(formals(colFun))]
  do.call("matrixToPng", c(list(targetMatrix, colFun = colFun), args[nms]))
}

handlerFunctionStars <- function() {
  args <- as.list(environment())[-(1:3)]
  targetGrid <- targetGrid(x, y, z, private$tileSize)
  gridDim <- which(names(st_dimensions(private$grid)) %in% attr(st_dimensions(private$grid), "raster")$dimensions)
  s <- lapply(mapply("==", lapply(
    lapply(names(st_dimensions(private$grid))[-(gridDim)], st_get_dimension_values, .x = private$grid),
    as.character
  ), args[names(st_dimensions(private$grid))[-(gridDim)]], SIMPLIFY = F), which)
  subGrid <- eval(expr(private$grid[attrib, , , !!!s]))
  d <- st_dimensions(subGrid) # check for curvilinear is same as in stars
  r <- attr(x, "raster")
  if (
    all(r$dimensions %in%
      names(x)) && isTRUE(attr(d, "raster")$curvilinear)) {
    warning("st_warp fails on curvilinear grids, trying slower fall back method")
    targetGridWithValues <- targetGrid
    a <- st_join(st_transform(st_as_sf(targetGrid,
      as_points = T
    ), st_crs(subGrid)), st_as_sf(subGrid,
      as_points = F
    ))
    targetGridWithValues$values <- (st_drop_geometry(a[, 2])[[1]])
  } else {
    targetGridWithValues <- st_warp(subGrid, targetGrid)
  }
  targetMatrix <- (drop(targetGridWithValues[[1]]))
  nms <- names(args)[names(args) %in% names(formals(colFun))]
  do.call("matrixToPng", c(list(targetMatrix, colFun = colFun), args[nms]))
}


inst/compose/script.R
require(stars)
require(starsTileServer)
s5p <- system.file(
  "sentinel5p/S5P_NRTI_L2__NO2____20180717T120113_20180717T120613_03932_01_010002_20180717T125231.nc",
  package = "starsdata"
)
nit <- read_stars(
  s5p,
  along = NA,
  sub = c(
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/nitrogendioxide_total_column",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/nitrogendioxide_total_column_precision",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/nitrogendioxide_total_column_precision_kernel",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/number_of_iterations",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/number_of_spectral_points_in_retrieval",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/oxygen_oxygen_dimer_slant_column_density",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/oxygen_oxygen_dimer_slant_column_density_precision",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/ozone_slant_column_density",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/ozone_slant_column_density_precision",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/processing_quality_flags",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/ring_coefficient",
    "//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/ring_coefficient_precision"
  ),
  curvilinear = c("//PRODUCT/longitude", "//PRODUCT/latitude"),
  driver = NULL
)
names(nit) <-
  sub("//PRODUCT/SUPPORT_DATA/DETAILED_RESULTS/", "", names(nit))
for (i in seq(length(names(nit)))) {
  nit[[i]][nit[[i]] > 9e+36] <- NA
}
st_crs(nit) <- 4326

server <- starsTileServer$new(nit)
# we save the server here as there should only be one version (sampling of color scales would otherwise result in differently colored tiles)
saveRDS(server, "server.rds")


R/starsTileServer.R

#' A R6 class that extends plumber to function as a tile server
#' @name starsTileServer
#'
#' @description
#' Creates a tile server based on the R6 class `plumber::Plumber`.
#' In can be created both with a `stars` grid as well as a function or list of functions.
#' The main methods are `run` and `new`.
#'
#' @examples
#' m <- matrix(1:20, nrow = 5, ncol = 4)
#' dim(m) <- c(x = 5, y = 4) # named dim
#' (s <- stars::st_as_stars(m))
#' sf::st_crs(s) <- 4326
#' starsTileServer$new(s)
#' # Working directly from a file
#' grid <- system.file("tif/L7_ETMs.tif", package = "stars")
#' starsTileServer$new(grid)
#' \dontrun{
#' starsTileServer$new(s)$run()
#' }
#' @export

starsTileServer <- R6Class(
  "starsTileServer",
  portable = T,
  private = list(
    grid = NULL,
    tileSize = NULL,
    staticParams = list(
      z = list(
        desc = "Zoom level (starts at 0)",
        type = "int",
        required = T
      ),
      x = list(
        desc = "Tile number in x direction (starts at 0)",
        type = "int",
        required = T
      ),
      y = list(
        desc = "Tile number in y direction (starts at 0)",
        type = "int",
        required = T
      )
    ),
    get_params = function() {
      a <-
        lapply(lapply(
          self$get_non_spatial_dimensions(),
          self$get_dimension_values_chr
        ), function(x) {
          if (length(x) == 1) {
            sprintf("[%s]", x)
          } else {
            if (length(x) == 2) {
              sprintf("[%s, %s]", x[1], x[2])
            } else {
              if (length(x) == 3) {
                sprintf("[%s, %s, %s]", x[1], x[2], tail(x, 1))
              } else {
                sprintf("[%s, %s, ..., %s]", x[1], x[2], tail(x, 1))
              }
            }
          }
        })
      if (length(a) != 0) {
        a <-
          lapply(paste0(self$get_non_spatial_dimensions(), ", values: ", a), function(x) {
            list(
              desc = x,
              type = "chr",
              required = F
            )
          })
        names(a) <- self$get_non_spatial_dimensions()
      }
      c(private$staticParams, a)
    }
  ),
  public = list(
    #' @description
    #' This method is used to initialize a new tile server
    #'
    #' @param grid Either a stars grid, a path pointing towards a gridded file, a function or a list of named functions
    #' @param colorFun a color function to use for coloring the map tiles, the function needs to be the same format as [leaflet::colorNumeric()]. It is important to specify a color function as it is important to keep the range of the color scale similar between tiles, therefore the minimum and maximum needs to be fixed. It can also be a list of color functions.
    #' @param tileSize The size of the tile (generally 256 pixels, and not tested with other sizes)
    #' @param ... Arguments passed on to the [plumber::Plumber], most important is the `port` number.
    #'
    #' @details
    #' If `grid` is a function it should take a stars grid as the first argument and return the grid with the same topology with values attached.
    #' Any other arguments to the function will be part of the API and will be passed to the function as characters.
    #'
    #' @return An `starsTileServer` object.
    initialize = function(grid,
                          colorFun = NULL,
                          tileSize = 256,
                          ...) {
      super$initialize(...)
      private$tileSize <- tileSize
      self$filter("cors", function(res) {
        # makes it possible cache tiles in browser
        res$setHeader("Access-Control-Allow-Origin", "*")
        plumber::forward()
      })
      # grid is a single function fall back on the list method
      if (is.function(grid)) {
        grid <- list(map = grid)
      }
      # grid is a list of functions in this case a method for each function is initiated
      if (is.list(grid) && !inherits(grid, "stars")) {
        stopifnot(all(unlist(lapply(
          grid, is.function
        ))))
        stopifnot(!is.null(names(grid)))
        stopifnot(all(names(grid) != ""))
        private$grid <- grid
        if (is.function(colorFun)) {
          colorFun <- rep(list(colorFun), length(names(private$grid)))
          names(colorFun) <- names(grid)
        }
        for (i in names(grid)) {
          stopifnot(i %in% names(colorFun))
          frms <- c(
            formals(grid[[i]])[-1],
            formals(colorFun[[i]])[-1]
          )
          formals(handlerFunctionFun) <-
            c(list(z = 0, x = 0, y = 0), frms)
          extraParams <-
            lapply(paste0(names(frms), ", values: [", unlist(frms), ", ...]"), function(x) {
              list(
                desc = x,
                type = "chr",
                required = F
              )
            })
          names(extraParams) <- names(frms)
          self$add_tile_endpoint(
            prefix = i,
            handlerFun = parse(text = sub(
              "idName", i, deparse(handlerFunctionFun)
            )),
            params = c(private$staticParams, extraParams),
            colorFun = colorFun[names(grid) == i][[1]]
          )
        }
      } else {
        # grid is a stars object add endpoints accordingly
        if (inherits(grid, "stars")) {
          # store the grid in the server as a stars object
          private$grid <- grid
        } else {
          if (is.character(grid)) {
            # if character consider it a path and read it
            stopifnot(file.exists(grid))
            private$grid <- read_stars(grid)
          } else {
            stop(
              "grid should be either, a function, list of functions, stars object or a path readable as a stars object with read_stars."
            )
          }
        }
        #  CRS is needed to locate the grids on the map
        stopifnot(!is.na(sf::st_crs(private$grid)))
        # upto here a grid has been read
        # Add a default color function
        if (is.null(colorFun)) {
          colorFun <-
            lapply(private$grid, function(y) {
              a <- structure(y, dim = NULL)
              rangeForColor <- range(
                sample((a), pmin(length(a), 1000)),
                na.rm = T
              )
              # Remove to prevent large returned objects
              rm(y, a)
              if (inherits(rangeForColor, "units")) {
                rangeForColor <- drop_units(rangeForColor)
              }
              f <- colorNumeric("viridis", rangeForColor)
              ff <- function(x, alpha = 1) {
                paste0(f(x), as.character(as.raw(
                  as.numeric(alpha) * 255
                )))
              }
              attributes(ff) <- attributes(f)
              ff
            })
        }
        # repeat the color function for each attribute of the stars object
        if (is.function(colorFun)) {
          colorFun <-
            rep(list(colorFun), length(names(private$grid)))
        }
        l <-
          lapply(self$get_non_spatial_dimensions(), function(x, g) {
            self$get_dimension_values_chr(x)[1]
          }, g = private$grid)
        names(l) <- self$get_non_spatial_dimensions()
        formals(handlerFunctionStars) <-
          c(list(z = 0, x = 0, y = 0), l, colFormals <- formals(colorFun[[1]])[-1])
        params <- private$get_params()
        if (length(colFormals) != 0) {
          extraParams <-
            lapply(paste0(names(colFormals), ", values: [", unlist(colFormals), ", ...]"), function(x) {
              list(
                desc = x,
                type = "chr",
                required = F
              )
            })
          names(extraParams) <- names(colFormals)
          params <- c(params, extraParams)
        } else {
          extraParams <- list()
        }
        # Loop over stars grid and add plumber end point for all of them
        for (i in names(private$grid)) {
          self$add_tile_endpoint(
            prefix = paste0("map/", i), colorFun = colorFun[[which(names(private$grid) == i)]],
            parse(text = deparse(handlerFunctionStars)), params = params
          )
        }
      }
    },
    #' @description
    #' Add three endpoints to the tile server, to return both the tiles and the color scale used.
    #'
    #' @param prefix the name to be used by the server for this tile server
    #' @param handlerFun The function that handles the api request and returns the grid
    #' @param colorFun The color function to use for example [leaflet::colorNumeric()]
    #' @param params parameters passed on to the `new` method of [plumber::PlumberEndpoint]
    #'
    add_tile_endpoint = function(prefix, handlerFun, colorFun, params) {
      env <- new.env()
      assign("attrib", basename(prefix), env)
      assign("colFun", colorFun, env)
      endpt <-
        plumber::PlumberEndpoint$new(
          "GET",
          paste0("/", prefix, "/<z:int>/<x:int>/<y:int>"),
          handlerFun,
          env,
          serializer = serializer_content_type("image/png"),
          params = params,
          comments = "This provide tiles."
        )
      self$handle(endpoint = endpt)
      endpt <-
        plumber::PlumberEndpoint$new(
          "GET",
          paste0("/", prefix, "/colorfunction"),
          function() {
            colorFun
          },
          new.env(),
          serializer = serializer_rds(),
          params = list(),
          comments = "An API end point to retrieve the color scale used by the server."
        )
      self$handle(endpoint = endpt)
      endpt <-
        plumber::PlumberEndpoint$new(
          "GET",
          paste0("/", prefix, "/colorfunctionnoalpha"),
          function() {
            a <- colorFun
            aa <- function(x) substr(a(x), 1, 7)
            attributes(aa) <- attributes(a)
            aa
          },
          new.env(),
          serializer = serializer_rds(),
          params = list(),
          comments = "An API end point to retrieve the color scale used by the server."
        )
      self$handle(endpoint = endpt)
    },
    #' @description
    #' return the grid used to initialize the function
    get_grid = function() {
      private$grid
    },
    #' @description
    #' return the attributes of the stars grid
    get_attributes = function() {
      names(private$grid)
    },
    #' @description
    #' return the names of the dimensions of the grid
    get_dimensions = function() {
      names(st_dimensions(private$grid))
    },
    #' @description
    #' return the values of a dimension as a character
    #'
    #' @param x the name of the dimension
    get_dimension_values_chr = function(x) {
      as.character(st_get_dimension_values(x, .x = private$grid))
    },
    #' @description
    #' return all non spatial dimensions
    get_non_spatial_dimensions = function() {
      self$get_dimensions()[!self$get_dimensions() %in% attr(st_dimensions(private$grid), "raster")$dimensions]
    }
  ),
  inherit = plumber::Plumber
)
NULL
